tools/api_proto_plugin/annotations.py

```diff
@@ -6,6 +6,9 @@
 
 # Key-value annotation regex.
 ANNOTATION_REGEX = re.compile('\[#([\w-]+?):(.*?)\]\s?', re.DOTALL)
+ANNOTATION_LINE_REGEX_FORMAT = '[^\S\r\n]*\[#%s:.*?\][^\S\r\n]*\n?'
+
+ANNOTATION_LINE_FORMAT = ' [#%s:%s]\n'
 
```

Comment:
Why is there now a distinction between line and non-line regexes? Can you add some comments if this is essential?

Reply:
I'm guessing this makes replacement operations easier or the like.

Reply:
It would leave an empty line in place with non-line regex. The full logical is that remove the annotation totally include the prefix whitespaces and suffix line separator. Then append annotations transformed and formatted to the end of leading comment.

Reply:
I still think there is a semantic gap here; we have a class today of regular annotations, these can appear anywhere. I _think_ that what you're doing is trying to add a separate class of annotations, "line annotations". These have to appear on a single line (on their own? only at the end of a comment block), to make it easier to substitute or replace them via single line regex primitives. Is my understanding here correct or off?

If it's correct, I think we need to explain in comments the above, and make clear which annotations are "line annotations" and which ones are not.

FWIW, I actually think this is a really good scheme, it simplifies a lot of things and many annotations don't need the flexibility to be anywhere. There are some that do, e.g. `[#comment:]`, but not `[#next-free-field:`.

Reply:
Thanks @htuch for leading me to think more.

First of all, this regex matches not only one single line, maybe it has a confusing name, it try to match many leading and trailing white spaces and a trailing line if present, which allow multiple lines.

Got your point if really :), definitely no annotation has the constraint that it has to appear on a single line and at the end of a comment block, especially like `[#comment:]`, and the transform for annotations should happen in place if no explicit requirement and not change the context of comment.

Then for those like `[#next-free-field:]`, i think it would be better to place at the end of comment on its own line. so we should specify it explicitly.

For now, i would make it simpler to complete this next free annotation task.